stan-dev/stan
# Feature/issue 1951 compound declaration definition

#### Submisison Checklist


#### Summary

Adds compound declaration/definition statements such as `real x = 1.0` to Stan language.

#### Intended Effect

For any variable that is assignable, allow assignment to be combined with the variable definition.
Assignable variable are local variables and variables declared in the transformed data, transformed parameters, and generated quantities blocks.

#### How to Verify

To verify that the parser correctly recognizes valid compound declaration/definition statements, run unit test `stan/src/test/unit/lang/parser/var_decls_grammar_test.cpp`.

To verify that the generated code compiles correctly, make the test models in `src/test/test-models/good/declare-define-var*.stan`,  e.g.:
```
for x in src/test/test-models/good/declare-define-*.stan
do
echo ${x%.stan}
make ${x%.stan}.hpp-test
done
```

#### Side Effects

The compiler cannot check that the dimension sizes match.  See example model: `src/test/test-models/good/declare-define-var-vec-runtime-error.stan`

#### Documentation

Basic syntax described in Stan manual.  More documentation is needed.  Future planned work is to update the example models to use this feature.

#### Reviewer Suggestions

Daniel Lee

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Mitzi Morris

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
